From 9744bee9c4f17ce6414c30039e9acfb2cbae38a3 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Thu, 14 Aug 2025 00:08:57 +0100
Subject: [PATCH 6/7] CertificateTool: port from PKCS#12 to BCFKS

Note this changes the keystore type inline without consulting the
default type. A more portable solution would be to use
KeyStore.getDefaultType().
---
 .../common/ssl/KeyStoreUtil.java              |  2 ++
 .../xpack/security/cli/AutoConfigureNode.java | 14 +++++------
 .../xpack/security/cli/CertGenUtils.java      |  6 ++---
 .../security/cli/CertificateGenerateTool.java |  4 +--
 .../xpack/security/cli/CertificateTool.java   | 25 ++++++++++---------
 .../security/cli/HttpCertificateCommand.java  | 12 ++++-----
 6 files changed, 31 insertions(+), 32 deletions(-)

diff --git a/libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/KeyStoreUtil.java b/libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/KeyStoreUtil.java
index 85bd5368b20..320a52a0253 100644
--- a/libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/KeyStoreUtil.java
+++ b/libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/KeyStoreUtil.java
@@ -55,6 +55,8 @@ public final class KeyStoreUtil {
         String name = path == null ? "" : path.toLowerCase(Locale.ROOT);
         if (name.endsWith(".p12") || name.endsWith(".pfx") || name.endsWith(".pkcs12")) {
             return "PKCS12";
+        } else if (name.endsWith(".bcfks")) {
+            return "BCFKS";
         } else {
             return "jks";
         }
diff --git a/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/AutoConfigureNode.java b/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/AutoConfigureNode.java
index 689134a5eba..ae5f028f4d7 100644
--- a/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/AutoConfigureNode.java
+++ b/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/AutoConfigureNode.java
@@ -543,7 +543,7 @@ public class AutoConfigureNode extends EnvironmentAwareCommand {
                 );
             }
             try (SecureString transportKeystorePassword = newKeystorePassword()) {
-                KeyStore transportKeystore = KeyStore.getInstance("PKCS12");
+                KeyStore transportKeystore = KeyStore.getInstance("BCFKS");
                 transportKeystore.load(null);
                 // the PKCS12 keystore and the contained private key use the same password
                 transportKeystore.setKeyEntry(
@@ -555,7 +555,7 @@ public class AutoConfigureNode extends EnvironmentAwareCommand {
                 transportKeystore.setCertificateEntry(TRANSPORT_CA_CERT_KEYSTORE_ENTRY, transportCaCert);
                 fullyWriteFile(
                     tempGeneratedTlsCertsDir,
-                    TRANSPORT_AUTOGENERATED_KEYSTORE_NAME + ".p12",
+                    TRANSPORT_AUTOGENERATED_KEYSTORE_NAME + ".bcfks",
                     false,
                     inReconfigureMode ? ELASTICSEARCH_GROUP_OWNER : null,
                     stream -> transportKeystore.store(stream, transportKeystorePassword.getChars())
@@ -565,7 +565,7 @@ public class AutoConfigureNode extends EnvironmentAwareCommand {
                 nodeKeystore.setString("xpack.security.transport.ssl.truststore.secure_password", transportKeystorePassword.getChars());
             }
             try (SecureString httpKeystorePassword = newKeystorePassword()) {
-                KeyStore httpKeystore = KeyStore.getInstance("PKCS12");
+                KeyStore httpKeystore = KeyStore.getInstance("BCFKS");
                 httpKeystore.load(null);
                 // the keystore contains both the node's and the CA's private keys
                 // both keys are encrypted using the same password as the PKCS12 keystore they're contained in
@@ -583,7 +583,7 @@ public class AutoConfigureNode extends EnvironmentAwareCommand {
                 );
                 fullyWriteFile(
                     tempGeneratedTlsCertsDir,
-                    HTTP_AUTOGENERATED_KEYSTORE_NAME + ".p12",
+                    HTTP_AUTOGENERATED_KEYSTORE_NAME + ".bcfks",
                     false,
                     inReconfigureMode ? ELASTICSEARCH_GROUP_OWNER : null,
                     stream -> httpKeystore.store(stream, httpKeystorePassword.getChars())
@@ -737,7 +737,7 @@ public class AutoConfigureNode extends EnvironmentAwareCommand {
                     bw.newLine();
                     bw.write("  enabled: true");
                     bw.newLine();
-                    bw.write("  keystore.path: " + TLS_GENERATED_CERTS_DIR_NAME + "/" + HTTP_AUTOGENERATED_KEYSTORE_NAME + ".p12");
+                    bw.write("  keystore.path: " + TLS_GENERATED_CERTS_DIR_NAME + "/" + HTTP_AUTOGENERATED_KEYSTORE_NAME + ".bcfks");
                     bw.newLine();
                     bw.newLine();
                     bw.write("# Enable encryption and mutual authentication between cluster nodes");
@@ -748,12 +748,12 @@ public class AutoConfigureNode extends EnvironmentAwareCommand {
                     bw.newLine();
                     bw.write("  verification_mode: certificate");
                     bw.newLine();
-                    bw.write("  keystore.path: " + TLS_GENERATED_CERTS_DIR_NAME + "/" + TRANSPORT_AUTOGENERATED_KEYSTORE_NAME + ".p12");
+                    bw.write("  keystore.path: " + TLS_GENERATED_CERTS_DIR_NAME + "/" + TRANSPORT_AUTOGENERATED_KEYSTORE_NAME + ".bcfks");
                     bw.newLine();
                     // we use the keystore as a truststore in order to minimize the number of auto-generated resources,
                     // and also because a single file is more idiomatic to the scheme of a shared secret between the cluster nodes
                     // no one should only need the TLS cert without the associated key for the transport layer
-                    bw.write("  truststore.path: " + TLS_GENERATED_CERTS_DIR_NAME + "/" + TRANSPORT_AUTOGENERATED_KEYSTORE_NAME + ".p12");
+                    bw.write("  truststore.path: " + TLS_GENERATED_CERTS_DIR_NAME + "/" + TRANSPORT_AUTOGENERATED_KEYSTORE_NAME + ".bcfks");
                     if (inEnrollmentMode) {
                         bw.newLine();
                         bw.write("# Discover existing nodes in the cluster");
diff --git a/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertGenUtils.java b/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertGenUtils.java
index b82e2e1d77f..4e48b570d09 100644
--- a/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertGenUtils.java
+++ b/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertGenUtils.java
@@ -27,7 +27,6 @@ import org.bouncycastle.cert.X509CertificateHolder;
 import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
 import org.bouncycastle.cert.jcajce.JcaX509ExtensionUtils;
 import org.bouncycastle.cert.jcajce.JcaX509v3CertificateBuilder;
-import org.bouncycastle.jce.provider.BouncyCastleProvider;
 import org.bouncycastle.operator.ContentSigner;
 import org.bouncycastle.operator.OperatorCreationException;
 import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
@@ -76,7 +75,6 @@ public class CertGenUtils {
     private static final String CN_OID = "2.5.4.3";
 
     private static final int SERIAL_BIT_LENGTH = 20 * 8;
-    private static final BouncyCastleProvider BC_PROV = new BouncyCastleProvider();
 
     /**
      * The mapping of key usage names to their corresponding integer values as defined in {@code KeyUsage} class.
@@ -274,7 +272,7 @@ public class CertGenUtils {
         PrivateKey signingKey = caPrivKey != null ? caPrivKey : keyPair.getPrivate();
         ContentSigner signer = new JcaContentSignerBuilder(
             (Strings.isNullOrEmpty(signatureAlgorithm)) ? getDefaultSignatureAlgorithm(signingKey) : signatureAlgorithm
-        ).setProvider(CertGenUtils.BC_PROV).build(signingKey);
+        ).setProvider("BCFIPS").build(signingKey);
         X509CertificateHolder certificateHolder = builder.build(signer);
         return new JcaX509CertificateConverter().getCertificate(certificateHolder);
     }
@@ -352,7 +350,7 @@ public class CertGenUtils {
             builder.addAttribute(PKCSObjectIdentifiers.pkcs_9_at_extensionRequest, extGen.generate());
         }
 
-        return builder.build(new JcaContentSignerBuilder("SHA256withRSA").setProvider(CertGenUtils.BC_PROV).build(keyPair.getPrivate()));
+        return builder.build(new JcaContentSignerBuilder("SHA256withRSA").setProvider("BCFIPS").build(keyPair.getPrivate()));
     }
 
     /**
diff --git a/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateGenerateTool.java b/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateGenerateTool.java
index a342e3cca3e..3e8ec75fb2e 100644
--- a/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateGenerateTool.java
+++ b/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateGenerateTool.java
@@ -13,7 +13,6 @@ import joptsimple.OptionSpec;
 import org.bouncycastle.asn1.DERIA5String;
 import org.bouncycastle.asn1.x509.GeneralName;
 import org.bouncycastle.asn1.x509.GeneralNames;
-import org.bouncycastle.jce.provider.BouncyCastleProvider;
 import org.bouncycastle.openssl.PEMEncryptor;
 import org.bouncycastle.openssl.jcajce.JcaPEMWriter;
 import org.bouncycastle.openssl.jcajce.JcePEMEncryptorBuilder;
@@ -89,7 +88,6 @@ class CertificateGenerateTool extends EnvironmentAwareCommand {
         "[a-zA-Z0-9!@#$%^&{}\\[\\]()_+\\-=,.~'` ]{1," + MAX_FILENAME_LENGTH + "}"
     );
     private static final int DEFAULT_KEY_SIZE = 2048;
-    private static final BouncyCastleProvider BC_PROV = new BouncyCastleProvider();
 
     /**
      * Wraps the certgen object parser.
@@ -535,7 +533,7 @@ class CertificateGenerateTool extends EnvironmentAwareCommand {
             outputStream.putNextEntry(new ZipEntry(caDirName + "ca.key"));
             if (info.password != null && info.password.length > 0) {
                 try {
-                    PEMEncryptor encryptor = new JcePEMEncryptorBuilder("DES-EDE3-CBC").setProvider(BC_PROV).build(info.password);
+                    PEMEncryptor encryptor = new JcePEMEncryptorBuilder("DES-EDE3-CBC").setProvider("BCFIPS").build(info.password);
                     pemWriter.writeObject(info.privateKey, encryptor);
                 } finally {
                     // we can safely nuke the password chars now
diff --git a/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java b/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java
index b64e2178627..ba694a559b3 100644
--- a/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java
+++ b/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertificateTool.java
@@ -16,7 +16,6 @@ import org.bouncycastle.asn1.DERIA5String;
 import org.bouncycastle.asn1.x509.GeneralName;
 import org.bouncycastle.asn1.x509.GeneralNames;
 import org.bouncycastle.asn1.x509.KeyUsage;
-import org.bouncycastle.jce.provider.BouncyCastleProvider;
 import org.bouncycastle.openssl.PEMEncryptor;
 import org.bouncycastle.openssl.jcajce.JcaPEMWriter;
 import org.bouncycastle.openssl.jcajce.JcePEMEncryptorBuilder;
@@ -94,8 +93,7 @@ class CertificateTool extends MultiCommand {
     private static final String DEFAULT_CSR_ZIP = "csr-bundle.zip";
     private static final String DEFAULT_CERT_ZIP = "certificate-bundle.zip";
     private static final String DEFAULT_CA_ZIP = "elastic-stack-ca.zip";
-    private static final String DEFAULT_CA_P12 = "elastic-stack-ca.p12";
-    private static final BouncyCastleProvider BC_PROV = new BouncyCastleProvider();
+    private static final String DEFAULT_CA_P12 = "elastic-stack-ca.bcfks";
 
     static final String DEFAULT_CERT_NAME = "instance";
 
@@ -587,7 +585,7 @@ class CertificateTool extends MultiCommand {
             char[] password,
             Terminal terminal
         ) throws Exception {
-            final KeyStore pkcs12 = KeyStore.getInstance("PKCS12");
+            final KeyStore pkcs12 = KeyStore.getInstance("BCFKS");
             pkcs12.load(null);
             withPassword(fileName, password, terminal, true, p12Password -> {
                 if (isAscii(p12Password)) {
@@ -598,7 +596,7 @@ class CertificateTool extends MultiCommand {
                     pkcs12.store(output, p12Password);
                     return null;
                 } else {
-                    throw new UserException(ExitCodes.CONFIG, "PKCS#12 passwords must be plain ASCII");
+                    throw new UserException(ExitCodes.CONFIG, "Keystore passwords must be plain ASCII");
                 }
             });
         }
@@ -796,9 +794,9 @@ class CertificateTool extends MultiCommand {
             if (writeZipFile) {
                 outputName = DEFAULT_CERT_ZIP;
             } else if (options.has(nameSpec)) {
-                outputName = nameSpec.value(options) + ".p12";
+                outputName = nameSpec.value(options) + ".bcfks";
             } else {
-                outputName = "elastic-certificates.p12";
+                outputName = "elastic-certificates.bcfks";
             }
             final Path output = resolveOutputPath(terminal, options, outputName);
 
@@ -822,13 +820,13 @@ class CertificateTool extends MultiCommand {
                     terminal.println("Each instance has a certificate and private key.");
                     filesDescription = "the certificate, key, and CA certificate";
                 } else {
-                    terminal.println("Each instance has a single PKCS#12 (.p12) file containing the instance");
+                    terminal.println("Each instance has a single BCFKS (.bcfks) file containing the instance");
                     terminal.println("certificate, instance private key and the CA certificate");
-                    filesDescription = "this '.p12' file";
+                    filesDescription = "this '.bcfks' file";
                 }
             } else {
                 terminal.println("This file is a self contained file and can be copied and used 'as is'");
-                filesDescription = "this '.p12' file";
+                filesDescription = "this '.bcfks' file";
             }
             terminal.println("For each Elastic product that you wish to configure, you should copy");
             terminal.println(filesDescription + " to the relevant configuration directory");
@@ -894,7 +892,7 @@ class CertificateTool extends MultiCommand {
                             // write private key
                             writePemPrivateKey(terminal, options, outputStream, pemWriter, entryBase + ".key", pair.key);
                         } else {
-                            final String fileName = entryBase + ".p12";
+                            final String fileName = entryBase + ".bcfks";
                             final char[] outputPassword = super.getOutputPassword(options);
                             outputStream.putNextEntry(new ZipEntry(fileName));
                             writePkcs12(
@@ -1064,7 +1062,10 @@ class CertificateTool extends MultiCommand {
     }
 
     static PEMEncryptor getEncrypter(char[] password) {
-        return new JcePEMEncryptorBuilder("AES-128-CBC").setProvider(BC_PROV).build(password);
+        // Example 65 onwards in https://downloads.bouncycastle.org/fips-java/docs/BCFipsIn100.pdf
+        // But likely using org.bouncycastle.util.io.pem.PemWriter
+        // TODO port to approved methods
+        return new JcePEMEncryptorBuilder("AES-128-CBC").setProvider("BCFIPS").build(password);
     }
 
     /**
diff --git a/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/HttpCertificateCommand.java b/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/HttpCertificateCommand.java
index 8c5dd0144ea..492f1dc5055 100644
--- a/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/HttpCertificateCommand.java
+++ b/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/HttpCertificateCommand.java
@@ -229,12 +229,12 @@ class HttpCertificateCommand extends EnvironmentAwareCommand {
             terminal.println("If you do not wish to use a password, simply press <enter> at the prompt below.");
             password = readPassword(terminal, "Provide a password for the private key: ", true);
         } else {
-            terminal.println("Your private key(s) will be stored in a PKCS#12 keystore file named \"http.p12\".");
+            terminal.println("Your private key(s) will be stored in a BCFKS keystore file named \"http.bcfks\".");
             terminal.println("This type of keystore is always password protected, but it is possible to use a");
             terminal.println("blank password.");
             terminal.println("");
             terminal.println("If you wish to use a blank password, simply press <enter> at the prompt below.");
-            password = readPassword(terminal, "Provide a password for the \"http.p12\" file: ", true);
+            password = readPassword(terminal, "Provide a password for the \"http.bcfks\" file: ", true);
         }
 
         printHeader("Where should we save the generated files?", terminal);
@@ -393,7 +393,7 @@ class HttpCertificateCommand extends EnvironmentAwareCommand {
                     Set.of(new ExtendedKeyUsage(KeyPurposeId.id_kp_serverAuth))
                 );
 
-                final String p12Name = "http.p12";
+                final String p12Name = "http.bcfks";
                 final String ymlFile = "sample-elasticsearch.yml";
                 final Map<String, String> substitutions = buildSubstitutions(
                     env,
@@ -421,7 +421,7 @@ class HttpCertificateCommand extends EnvironmentAwareCommand {
                     env,
                     Map.of(
                         "P12",
-                        "ca.p12",
+                        "ca.bcfks",
                         "DN",
                         ca.certAndKey.cert.getSubjectX500Principal().getName(),
                         "PASSWORD",
@@ -429,10 +429,10 @@ class HttpCertificateCommand extends EnvironmentAwareCommand {
                     )
                 )
             );
-            final KeyStore pkcs12 = KeyStore.getInstance("PKCS12");
+            final KeyStore pkcs12 = KeyStore.getInstance("BCFKS");
             pkcs12.load(null);
             pkcs12.setKeyEntry("ca", ca.certAndKey.key, ca.password, new Certificate[] { ca.certAndKey.cert });
-            try (ZipEntryStream entry = new ZipEntryStream(zip, dirName + "/ca.p12")) {
+            try (ZipEntryStream entry = new ZipEntryStream(zip, dirName + "/ca.bcfks")) {
                 pkcs12.store(entry, ca.password);
             }
         } catch (KeyStoreException | IOException | CertificateException | NoSuchAlgorithmException e) {
-- 
2.48.1

